Update /instrument flag values

drognanar opened the pull request:
Make instrument

tmat commented on src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs:
```diff
@@ -109,7 +109,7 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
             bool errorEndLocation = false;
             bool reportAnalyzer = false;
-            string instrument = "";
+            InstrumentationKind instrument = InstrumentationKind.None;
```
var

tmat commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -32,10 +32,9 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
+        /// Type of instrumentation that should be added to the output binary.
         /// </summary>
-        public string Instrument { get; private set; }
+        public InstrumentationKind Instrument { get; private set; }
```
I think the property should be named "InstrumentationKind". It's odd to use a verb.

tmat commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -377,7 +379,7 @@ public EmitOptions WithIncludePrivateMembers(bool value)
         }
 
-        public EmitOptions WithInstrument(string instrument)
+        public EmitOptions WithInstrument(InstrumentationKind instrument)
```
```WithInstrumentation``` or ```WithInstrumentationKind```

JohnHamby commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -4428,8 +4428,8 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
- /instrument:options           Sets instrumentation options.
-                               If empty instrumentation is disabled.
+ /instrument:{TestCoverage}    Specify instrumentation method.
+                               If flag is not passed then instrumentation is disabled.
```
I'm not sure precisely what this means. Is "flag" "instrument" or "TestCoverage"?

I don't know if there is a way to word this such that adding a new kind of instrumentation involves only adding text.

JohnHamby commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -4936,4 +4936,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+  </data>
+  <data name="ERR_InvalidInstrumentationKind" xml:space="preserve">
+    <value>Invalid instrumentation kind: {0}</value>
+  </data>
```
Are both of the above necessary?

JohnHamby commented on src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs:
```diff
@@ -109,7 +109,7 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
             bool errorEndLocation = false;
             bool reportAnalyzer = false;
-            string instrument = "";
+            HashSet<string> instrument = new HashSet<string>();
```
Should the hash set be configured to do case-insensitive comparison?

JohnHamby commented:
At a quick glance this looks fine to me, modulo the couple of questions I asked (which I didn't repeat for the corresponding VB code).

tmat commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -32,10 +33,9 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
+        /// Type of instrumentation that should be added to the output binary.
         /// </summary>
-        public string Instrument { get; private set; }
+        public HashSet<string> Instrument { get; private set; }
```
>HashSet<string> [](start = 15, length = 15)

Why not enum flags?

tmat replied:
This is certainly wrong since HashSet is mutable. I'd prefer not use strings. If we use strings then ImmutableArray<string> would be more appropriate.

---
In reply to: [73979817](https://github.com/dotnet/roslyn/pull/13013#discussion_r73979817) [](ancestors = 73979817)

JohnHamby replied:
A flags enum is limited to 32 (or 64) values. An immutable array of non-flags enum values probably works.

tmat replied:
Are we gonna have more than 64 kinds of instrumentation? If that ever happens (I doubt it will) we can add another enum, or even immutable array later.

drognanar commented:
FYI @dotnet/roslyn-compiler

jcouv commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1426,5 +1426,13 @@ internal enum ErrorCode
+
+        #region diagnostics for instrumentation
+
+        ERR_InvalidInstrumentationKind = 8501,
```
Do you expect to have bunch of new error codes for instrumentation?
If not, we just compacted the C#7 error codes, so we'd rather not introduce new large gaps. You could grab `8111`.

JohnHamby replied:
@drognanar did this. (I'll be adding notes to myself so that I can figure out what remains to be done.)

jcouv commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -177,6 +178,14 @@ public bool Equals(EmitOptions other)
+            {
+                if (this.InstrumentationKinds[i] != other.InstrumentationKinds[i])
+                {
+                    return false;
```
It looks like order matters in this equality comparison. Can you confirm this is intentional?

JohnHamby replied:
@drognanar addressed this by using SequenceEqual, which does not require sequences to have equal values in the same order in order test as equal.

JohnHamby replied:
The code looks different now, because @drognanar changed it to use SequenceEqual, but equality of EmitOptions still implies identical order of instrumentation kinds. That isn't too significant now, in that only one instrumentation kind can be specified, but seems questionable. Tagging @gafter @AlekseyTs @cston for opinions.

jcouv commented on src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs:
```diff
@@ -111,7 +112,7 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
             bool errorEndLocation = false;
             bool reportAnalyzer = false;
-            string instrument = "";
+            ArrayBuilder<InstrumentationKind> instrument = ArrayBuilder<InstrumentationKind>.GetInstance();
```
In the past John said there would be an unbounded number of such kinds, but I still don't understand the concept of different instrumentation kinds.
Would you have an example of two different "kinds" (to clarify the concept) which can be simultaneously be instrumented (to justify why we need an array of them).

jcouv commented on src/Compilers/Core/Portable/Emit/InstrumentationKind.cs:
```diff
@@ -0,0 +1,20 @@
+    public enum InstrumentationKind
+    {
+        None = 0,
+        TestCoverage = 1,
```
Previously, @JohnHamby had indicated that instrumentation kinds would not be an enum. I'm confused.

JohnHamby replied:
I don't have a strong opinion about whether instrumentation kinds are encoded using an enum or as strings--whatever gets through review is fine with me.

An advantage of an enum is that the allowed values are explicit, and that's my preference, but mostly we need something that works merged as soon as practical.

jcouv commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -225,6 +239,14 @@ internal void ValidateOptions(DiagnosticBag diagnostics, CommonMessageProvider m
+            {
+                if (!instrumentationKind.IsValid())
+                {
+                    diagnostics.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_InvalidInstrumentationKind, Location.None, (int)instrumentationKind));
```
Clarifying the concept of instrumentation kind will also help me understand if/whether a numeric index is a meaningful way to represent it and communicate it to users. For instance, should an instrumentation kind have a number or an identifier/name?

Will this error message be useful? "Invalid instrumentation kind: 42"

tmat replied:
I think this is ok, since the user needs to cast int to the enum type to produce invalid value. All valid values are named (have a enum member). Invalid values are unnamed (don't have enum value), so there is nothing better to display.

jcouv commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -245,7 +267,7 @@ internal void ValidateOptions(DiagnosticBag diagnostics, CommonMessageProvider m
         }
 
-        internal bool EmitDynamicAnalysisData => !string.IsNullOrEmpty(Instrument);
+        internal bool EmitDynamicAnalysisData => !InstrumentationKinds.IsDefaultOrEmpty && InstrumentationKinds.Contains(InstrumentationKind.TestCoverage);
```
If I understood correctly, it seems that `InstrumentationKinds` cannot be default (it could either be empty or be filled).

JohnHamby replied:
@drognanar addressed this issue.

jcouv commented on src/Compilers/VisualBasic/Portable/Errors/MessageProvider.vb:
```diff
@@ -272,6 +272,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                Return ERRID.ERR_InvalidInstrumentationKind
+            End Get
+        End Property
+
```
Nit: extra empty line here

JohnHamby replied:
@drognanar addressed this issue.

jcouv commented on src/Compilers/Core/Portable/Emit/InstrumentationKind.cs:
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
```
The `using` is probably not required.

JohnHamby replied:
@drognanar removed the using.

jcouv commented:
Done with review pass. No major concern, mostly some clarifying design questions.

jcouv commented:
LGTM

ManishJayaswal commented:
@AlekseyTs for CR signoff.

gafter commented on src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs:
```diff
@@ -73,7 +73,7 @@ public string ChecksumAlgorithm
         
         /// <summary>
-        /// An instrument flag that specifies instrumentation settings.
+        /// An instrument flag that specifies instrumentation kind.
```
It is hard to tell whether this is intended to support a single kind of instrumentation, or multiple kinds. I does not help that this sentence is missing the article "the" (e.g. "the instrumentation kind") and also "kind" is singular, making it somewhat awkward and ambiguous.

JohnHamby replied:
@drognanar addressed this, clarifying that multiple instrumentation kinds can appear.

gafter commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -377,14 +399,19 @@ public EmitOptions WithIncludePrivateMembers(bool value)
+            return WithInstrument(ImmutableArray.Create(instrument));
+        }
+
+        public EmitOptions WithInstrument(ImmutableArray<InstrumentationKind> instrument)
```
Should be WithInstrumentationKinds.

AlekseyTs replied:
+1

---
In reply to: [75935486](https://github.com/dotnet/roslyn/pull/13013#discussion_r75935486) [](ancestors = 75935486)

JohnHamby replied:
@drognanar made the change to WithInstrumentationKinds.

gafter commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -377,14 +399,19 @@ public EmitOptions WithIncludePrivateMembers(bool value)
         }
 
-        public EmitOptions WithInstrument(string instrument)
+        public EmitOptions WithInstrument(params InstrumentationKind[] instrument)
```
>WithInstrument [](start = 27, length = 14)

Please remove this from the API. We generally have only the single overload with an immutable input for the With methods.

JohnHamby replied:
@drognanar removed the params variant.

gafter commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -133,7 +134,7 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
             bool tolerateErrors = false,
             bool includePrivateMembers = false,
-            string instrument = "")
+            ImmutableArray<InstrumentationKind> instrument = default(ImmutableArray<InstrumentationKind>))
```
>instrument [](start = 48, length = 10)

Should be called `instrumentationKinds`.

JohnHamby replied:
@drognanar made the requested name change.

gafter commented:
I'm done reviewing. I just have a couple of public API changes, please.

AlekseyTs commented on src/Compilers/Core/Portable/Emit/InstrumentationKind.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Microsoft.CodeAnalysis.Emit
+{
+    public enum InstrumentationKind
```
>InstrumentationKind [](start = 16, length = 19)

Please add doc comments to the type and its members.

JohnHamby replied:
@drognanar added doc comments.

AlekseyTs commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -32,10 +34,9 @@ public sealed class EmitOptions : IEquatable<EmitOptions>
+        /// Type of instrumentation that should be added to the output binary.
         /// </summary>
-        public string Instrument { get; private set; }
+        public ImmutableArray<InstrumentationKind> InstrumentationKinds { get; private set; }
```
>ImmutableArray<InstrumentationKind> [](start = 15, length = 35)

Could this just be a bitmask or do we expect to have more than 32/64 different kinds?

JohnHamby replied:
We are leaving open the possibility of having more than 64 different kinds.

AlekseyTs commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -189,11 +198,17 @@ public bool Equals(EmitOptions other)
                 this.TolerateErrors == other.TolerateErrors &&
                 this.IncludePrivateMembers == other.IncludePrivateMembers &&
-                this.Instrument == other.Instrument;
+                this.InstrumentationKinds == other.InstrumentationKinds;
```
>this.InstrumentationKinds == other.InstrumentationKinds [](start = 16, length = 55)

What does this comparison compare and do we want to do this comparison?

JohnHamby replied:
@drognanar changed the comparison to use SequenceEqual, which is more meaningful here. Therer is still a question about whether equality should allow differences in order of instrumentation kinds.

AlekseyTs commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -177,6 +178,14 @@ public bool Equals(EmitOptions other)
                 return false;
             }
 
+            for (int i = 0; i < this.InstrumentationKinds.Length; i++)
```
>i < this.InstrumentationKinds.Length [](start = 28, length = 36)

How are we handling different length between this.InstrumentationKinds and other.InstrumentationKinds?

JohnHamby replied:
@drognanar addressed this by using SequenceEqual.

AlekseyTs commented:
>         public bool Equals(EmitOptions other)

Do we have tests for Equals/GetHashCode with respect to InstrumentationKinds values?

---
Refers to: src/Compilers/Core/Portable/Emit/EmitOptions.cs:174 in 27c2a2f. [](commit_id = 27c2a2fa7c3e55f3c5f4631e713bf0ff58af444d, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -189,11 +198,17 @@ public bool Equals(EmitOptions other)
+            foreach (var instrumentationKind in InstrumentationKinds)
+            {
+                instrumentationKindHash = Hash.Combine((int)instrumentationKind, instrumentationKindHash);
+            }
```
Note, we have Hash.CombineValues and SequenceEquals helpers, see how they are used in ParseOptions.

JohnHamby replied:
The foreach loop will be removed.

AlekseyTs commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -245,7 +267,7 @@ internal void ValidateOptions(DiagnosticBag diagnostics, CommonMessageProvider m
         }
 
-        internal bool EmitDynamicAnalysisData => !string.IsNullOrEmpty(Instrument);
+        internal bool EmitDynamicAnalysisData => InstrumentationKinds.Contains(InstrumentationKind.TestCoverage);
```
>EmitDynamicAnalysisData [](start = 22, length = 23)

Perhaps the property should be renamed to EmitTestCoverageData?

JohnHamby replied:
@drognanar made the requested name change.

AlekseyTs commented:
>         public void TestFieldsForEqualsAndGetHashCode()

I think we need more targeted tests for equality around InstrumentationKinds.

---
Refers to: src/Compilers/Core/CodeAnalysisTest/Emit/EmitOptionsTests.cs:65 in 27c2a2f. [](commit_id = 27c2a2fa7c3e55f3c5f4631e713bf0ff58af444d, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/MSBuildTask/Shared/ManagedCompiler.cs:
```diff
@@ -73,7 +73,7 @@ public string ChecksumAlgorithm
-        /// An instrument flag that specifies instrumentation settings.
+        /// An instrument flag that specifies instrumentation kind.
         /// </summary>
         public string Instrument
```
>Instrument [](start = 22, length = 10)

Should this property be renamed? Also, what format the value is using?

AlekseyTs commented on src/Compilers/Core/MSBuildTaskTests/CscTests.cs:
```diff
@@ -113,8 +113,12 @@ public void ChecksumAlgorithmOption()
         public void InstrumentTestNamesFlag()
```
>InstrumentTestNamesFlag [](start = 20, length = 23)

How does one specify more than one kind?

JohnHamby replied:
Adding a test that specifies multiple kinds exposed a bug in how the option was generated. Fixed in the next commit.

AlekseyTs commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -2996,39 +2996,53 @@ public void ParseOut2()
         {
             var parsedArgs = DefaultParse(SpecializedCollections.EmptyEnumerable<string>(), _baseDirectory);
-            Assert.Equal("", parsedArgs.EmitOptions.Instrument);
+            Assert.Equal(ImmutableHashSet<InstrumentationKind>.Empty, parsedArgs.EmitOptions.InstrumentationKinds);
```
>ImmutableHashSet [](start = 25, length = 16)

Is the property an ImmutableArray or an ImmutableHashSet?

JohnHamby replied:
@drognanar restructured the tests in a way that resolves this issue.

AlekseyTs commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -2996,39 +2996,53 @@ public void ParseOut2()
         public void ParseInstrumentTestNames()
```
>ParseInstrumentTestNames [](start = 20, length = 24)

It looks like we are not testing scenarios when /instrument switch is specified multiple times.

AlekseyTs commented on src/Dependencies/PooledObjects/Microsoft.CodeAnalysis.PooledObjects.shproj:
```diff
@@ -1,7 +1,7 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup Label="Globals">
-    <ProjectGuid>c1930979-c824-496b-a630-70f5369a636f</ProjectGuid>
+    <ProjectGuid>cd77a8cc-2885-47a7-b99c-fbf13353400b</ProjectGuid>
```
><ProjectGuid>cd77a8cc-2885-47a7-b99c-fbf13353400b</ProjectGuid> [](start = 4, length = 63)

Is this change necessary?

AlekseyTs commented:
I am done with review pass.

drognanar commented:
Updated PR.
Added a test validating GetHash,Equals for InstrumetnationKinds property.
Added a test that parses multiple instrumentation kinds (one of the strings is invalid since there only exists one InstrumentationKind).

cston commented on src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs:
```diff
@@ -1656,6 +1657,24 @@ private static void ValidateWin32Settings(string win32ResourceFile, string win32
             }
         }
 
+        internal static IEnumerable<InstrumentationKind> ParseInstrumentationKinds(string value, IList<Diagnostic> diagnostics)
```
`private`

cston replied:
Consider passing in the `ArrayBuilder<InstrumentationKind>` rather than returning `IEnumerable<InstrumentationKind>`. That would avoid an iterator class and would be consistent with `diagnostics` which is an argument.

cston replied:
No need to change from iterator. It looks like that is the pattern for `Parse` methods.

cston replied:
Not part of this change but it looks like `ParseAnalyzers` and `ParseAssemblyReferences` can be `static`. Same for `ParseAnalyzers` in VB.

AlekseyTs commented on src/Compilers/Core/MSBuildTask/Shared/Microsoft.CSharp.Core.targets:
```diff
@@ -95,7 +95,7 @@
           GenerateFullPaths="$(GenerateFullPaths)"
           HighEntropyVA="$(HighEntropyVA)"
-          Instrument="$(Instrument)"
+          InstrumentationKinds="$(Instrument)"
```
>Instrument [](start = 34, length = 10)

Should this be renamed too? Also, it looks like things that might have multiple values are using Items rather than a property.

JohnHamby replied:
After lots of back and forth, the current deliberate decision is that Instrument is a property. There is precedent for multiple items being combined in a property, e.g. NoWarn.

Also, "InstrumentKinds" above will revert to "Instrument" to match the spelling of the command-line option.

AlekseyTs commented on src/Compilers/Core/Portable/Emit/InstrumentationKind.cs:
```diff
@@ -0,0 +1,28 @@
+        None = 0,
+
+        /// <summary>
+        /// Instruments the binary to add....
```
>Instruments the binary to add.... [](start = 12, length = 33)

It looks like the comment is incomplete.

JohnHamby replied:
I think the intent was that the name appearing in the code complete the comment. Since this has confused at least one reviewer, I'll make the comment self contained, with apologies to @drognanar .

AlekseyTs commented:
>         Public Sub ParseInstrumentTestNames()

It looks like we are missing a test with multiple valid values, even duplicate ones. Same true for C#.

---
Refers to: src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb:421 in 74a7879. [](commit_id = 74a78795f598c7e248ff732d9b3b34f9a44b5bd4, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Emit/EmitOptions.cs:
```diff
@@ -189,11 +190,17 @@ public bool Equals(EmitOptions other)
+            foreach (var instrumentationKind in InstrumentationKinds)
+            {
+                instrumentationKindHash = Hash.Combine((int)instrumentationKind, instrumentationKindHash);
+            }
```
Do we still need this code?

JohnHamby replied:
No, this code will be removed.

AlekseyTs commented:
Done with review pass.

JohnHamby commented:
I will add tests for multiple valid duplicate values for CSharp and for VB.
